### PR TITLE
(vertical compaction) fix vertical compaction core

### LINKED ISSUE #22139: be进程崩溃
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

doris-2.0-beta-afe6bb9

### What's Wrong?

全部be进程都消失，be的日志看到错误内容：
F0723 17:51:25.067346 58823 segment_writer.cpp:746] Check failed: _row_count == _num_rows_written (50184 vs. 50185) 
*** Check failure stack trace: ***
    @     0x7f8d00175619  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f8cf91c38ae  doris::segment_v2::SegmentWriter::finalize_columns_data()
    @     0x7f8cf921aa81  doris::VerticalBetaRowsetWriter::_flush_columns()
    @     0x7f8cf921ae59  doris::VerticalBetaRowsetWriter::flush_columns()
    @     0x7f8cf8b60315  doris::Merger::vertical_compact_one_group()
    @     0x7f8cf8b61306  doris::Merger::vertical_merge_rowsets()
    @     0x7f8cf8b52a10  doris::Compaction::do_compaction_impl()
    @     0x7f8cf8b51e3a  doris::Compaction::do_compaction()
    @     0x7f8cf92d924c  doris::BaseCompaction::execute_compact_impl()
    @     0x7f8cf8b51b13  doris::Compaction::execute_compact()
    @     0x7f8cf92fcbce  doris::Tablet::execute_compaction()
    @     0x7f8cf8af39fe  std::_Function_handler<>::_M_invoke()
    @     0x7f8cf96230bf  doris::ThreadPool::dispatch_thread()
    @     0x7f8cf9618f5c  doris::Thread::supervise_thread()
    @     0x7f8ced1f8ea5  start_thread
    @     0x7f8ced1f8ea5  start_thread
    @     0x7f8ceda11b0d  __clone

### What You Expected?

be能稳定运行

### How to Reproduce?

_No response_

### Anything Else?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by hust-hhb**
co-author:@zhannngchen

## Proposed changes

Issue Number: close #22139

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hust-hhb** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.59 seconds
 stream load tsv:          506 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.5 seconds inserted 10000000 Rows, about 338K ops/s
 storage size: 17170954103 Bytes

**Comment by hust-hhb** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.73 seconds
 stream load tsv:          509 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162502620 Bytes

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 1]
```diff
@@ -631,7 +631,7 @@ Status SegmentWriter::append_block(const vectorized::Block* block, size_t row_po
         if (_tablet_schema->keys_type() == UNIQUE_KEYS && _opts.enable_unique_key_merge_on_write) {
             // create primary indexes
             std::string last_key;
-            for (size_t pos = 0; pos < num_rows; pos++) {
+            for (size_t pos = row_pos; pos < num_rows; pos++) {
```
Comment (zhannngchen) [inline be/src/olap/rowset/segment_v2/segment_writer.cpp:634]: should be `pos < row_pos + num_rows`?

File: be/src/olap/rowset/vertical_beta_rowset_writer.cpp [hunk 2]
```diff
@@ -99,19 +99,53 @@ Status VerticalBetaRowsetWriter::add_columns(const vectorized::Block* block,
...
                     << ", _cur_writer_idx: " << _cur_writer_idx;
+        uint32_t num_rows_key_group = _segment_writers[_cur_writer_idx]->row_count();
+        DCHECK(num_rows_key_group > 0);
         // init if it's first value column write in current segment
         if (_cur_writer_idx == 0 && num_rows_written == 0) {
             VLOG_NOTICE << "init first value column segment writer";
             RETURN_IF_ERROR(_segment_writers[_cur_writer_idx]->init(col_ids, is_key));
         }
-        if (num_rows_written > max_rows_per_segment) {
+        size_t start_offset = 0, limit = num_rows;
```
Comment (zhannngchen) [inline be/src/olap/rowset/vertical_beta_rowset_writer.cpp:109]: Add more comments here

**Comment by hust-hhb** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.01 seconds
 stream load tsv:          529 seconds loaded 74807831229 Bytes, about 134 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17162972257 Bytes

**Comment by hust-hhb** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.05 seconds
 stream load tsv:          540 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17161240487 Bytes

**Comment by hust-hhb** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.49 seconds
 stream load tsv:          507 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.5 seconds inserted 10000000 Rows, about 338K ops/s
 storage size: 17170358852 Bytes

**Review by zhannngchen — APPROVED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**